### ISSUE TITLE
Error when generating/serializing keys with multilines and colon

### ISSUE THREAD

**Issue by eginez** [FasterXML/jackson-dataformats-text#306]
Hello there

I wanted to report that a key containing a colon `:` followed by new line chars `\n` does not seem to be parsing correctly. For example the following test case reproduces the error

```
    @Test(expected = JsonParseException.class)
    public void testYamlMultiLineNotCanonical() throws Exception {
        String key = "SomeKey:\n\nOtherLine";
        ObjectMapper jsonMapper = new ObjectMapper();
        ObjectNode jsonRoot = jsonMapper.createObjectNode();
        jsonRoot.put(key, 302);

        YAMLFactory yamlFactory = new YAMLFactory();
        ObjectMapper yamlObjectMapper = new ObjectMapper(yamlFactory);
        yamlObjectMapper.configure(JsonParser.Feature.STRICT_DUPLICATE_DETECTION, true);
        StringWriter swriter = new StringWriter();

        yamlObjectMapper.writeValue(swriter, jsonRoot);

        ObjectMapper readMapper =  new ObjectMapper(yamlFactory);
        ObjectNode readJsonRoot = (ObjectNode) readMapper.readTree(swriter.toString());
        Assert.assertEquals(readJsonRoot.get(key).asInt(),302 );
    }
```

I would have expected the parser to quote the key as `\n` and `:` are supported yaml content.

There is  work around for this:  Force canonical output.

```
        yamlFactory.configure(YAMLGenerator.Feature.CANONICAL_OUTPUT, true);
```

That however could have other side effects. Maybe I am missing some other configuration that would do the same?

Thoughts, comments are welcomed.

**Comment by cowtowncoder** [comment]
Sounds reasonable that such characters (if and when allowed by YAML) should be properly escaped so that round-trip handling would work. So sounds like a bug.
I don't remember if there are relevant settings, off the top of my head, but I guess there aren't (you probably checked :) ).

**Comment by eginez** [comment]
Thanks for confirming @cowtowncoder. As far as I can tell this bug is in SnakeYAML? Would that be a fair assessment?

**Comment by cowtowncoder** [comment]
@eginez It does sound like it would be, but I have not had a chance to check this yet. If you can write a simple reproduction against `SnakeYAML`, that'd allow filing a bug against it.

@asomov would be a good contact here, I think.

**Comment by asomov** [comment]
@eginez what is `swriter.toString()` ?

**Comment by asomov** [comment]
@eginez  @cowtowncoder I made a test and it is green:
```
  public void testKey() {
    String key = "SomeKey:\n\nOtherLine";
    Map<String, Integer> map = new HashMap<>();
    map.put(key, 302);
    Yaml yaml = new Yaml();
    String obj = yaml.dump(map);
    assertEquals("? |-\n"
        + "  SomeKey:\n"
        + "\n"
        + "  OtherLine\n"
        + ": 302\n", obj);
    Map<String, String> parsed = yaml.load(obj);
    assertEquals(302, parsed.get(key));
  }
```

**Comment by eginez** [comment]
@asomov interesting so the bug could indeed somewhere in the `jackson` library. We should expect this to work right???

```
@Test
    public void testYaml2() throws Exception {
        String key = "SomeKey:\n\nOtherLine";
        Map<String, Integer> map = new HashMap<>();
        map.put(key, 302);

        Yaml yaml = new Yaml();
        String dump = yaml.dump(map);

        YAMLFactory yamlFactory = new YAMLFactory();
        ObjectMapper yamlObjectMapper = new ObjectMapper(yamlFactory);
        StringWriter swriter = new StringWriter();
        yamlObjectMapper.writeValue(swriter, map);

        Assert.assertEquals(dump, swriter.toString());
    }
```

The keys difference is the absence of `|-` 

@cowtowncoder `swriter.toString()` pushes the contents of the string buffer to a string object

**Comment by cowtowncoder** [comment]
It does sound like potential Jackson bug. I can probably use/repurpose the original sample; the goal is not so much to replicate what specific SnakeYAML settings produce but to ensure that what is written is read back reliably.

I suspect that Jackson's logic that tries to prevent coercion of "NO" into `false` is leading it to specify some output mode that disables SnakeYAML's escaping or checks.

**Comment by eginez** [comment]
this is great thanks all!!

**Comment by asomov** [comment]
@cowtowncoder feel free to ping me if you need any help

**Comment by cowtowncoder** [comment]
@asomov Will do if need be, thanks!

Fix is in, for 2.13.2 release. Covers linefeed case but I suspect there may be other things that require quoting.
